erpnext/maintenance/doctype/maintenance_schedule/maintenance_schedule.js

```diff
@@ -152,12 +152,27 @@ erpnext.maintenance.MaintenanceSchedule = class MaintenanceSchedule extends frap
 		this.set_no_of_visits(doc, cdt, cdn);
 	}
 
+	serial_no(doc) {
+		let me = this;
+		if (doc.docstatus == 0) {
+			me.frm.call('generate_schedule');
+		}
+	}
+
+	sales_person(doc) {
+		let me = this;
+		if (doc.docstatus == 0) {
+			me.frm.call('generate_schedule');
+		}
+	}
+
```

Comment:
instead of calling `generate_schedule`  conditionally from JS and Py, call it purely from server side ?
Create one function on the py side that checks if generate schedule needs to be called and calls it on `validate`. The JS file needs a clean up, lets not add more to it maybe.

Probably first map out under what circumstances you want to regenerate/generate schedule in `validate` method

Reply:
Basically I need to call it whenever there's a change in the items table, how do I do that from server side. I can handle everything except the date change.